servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -251,52 +329,73 @@ public void onComplete() {
-                return succeeded(connection);
+            // don't open new connections for expired hosts, try a different one
+            if (host.isActive()) {
+                pickedHost = host;
+                break;
             }
         }
-
+        if (pickedHost == null) {
+            return failed(StacklessNoAvailableHostException.newInstance(
+                    "Failed to pick an active host. Either all are busy or all are expired.",
+                    RoundRobinLoadBalancer.class, "selectConnection0(...)"));
+        }
         // No connection was selected: create a new one.
+        final Host<ResolvedAddress, C> host = pickedHost;
         // This LB implementation does not automatically provide TransportObserver. Therefore, we pass "null" here.
         // Users can apply a ConnectionFactoryFilter if they need to override this "null" value with TransportObserver.
         return connectionFactory.newConnection(host.address, null)
                 .flatMap(newCnx -> {
-                    // Invoke the selector before adding the connection to the pool, otherwise, connection can be used
-                    // concurrently and hence a new connection can be rejected by the selector.
+                    // Invoke the selector before adding the connection to the pool, otherwise, connection can be
+                    // used concurrently and hence a new connection can be rejected by the selector.
                     if (!selector.test(newCnx)) {
-                        // Failure in selection could be temporary, hence add it to the queue and be consistent with the
-                        // fact that select failure does not close a connection.
+                        // Failure in selection could be temporary, hence add it to the queue and be consistent
+                        // with the fact that select failure does not close a connection.
                         return newCnx.closeAsync().concat(failed(new ConnectionRejectedException(
                                 "Newly created connection " + newCnx + " rejected by the selection filter.")));
                     }
                     if (host.addConnection(newCnx)) {
                         return succeeded(newCnx);
                     }
-                    return newCnx.closeAsync().concat(this.activeHosts == CLOSED_LIST ? failedLBClosed() :
-                            failed(new ConnectionRejectedException("Failed to add newly created connection for host: " +
-                                    host.address + ", host inactive? " + host.isInactive())));
+                    return newCnx.closeAsync().concat(this.usedHosts == CLOSED_LIST ? failedLBClosed() :
+                            failed(new ConnectionRejectedException(
+                                    "Failed to add newly created connection for host: "
+                                            + host.address
+                                            + ", host active? " + host.isActive()
```

Comment:
Can you use `Host#toString()` here? Then you could make `isClosed()` and `isExpired()` private.